Rewrite ordering documentation

pmatilai opened the pull request:
News about rpm 4.0.1 ordering algorithm changes are getting a bit
long in the tooth. Rewrite the whole doc to make it reflect modern rpm
behavior.

pmatilai commented:
(This addresses the last bullet from #1952)

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+
+`Requires` are the primary data for ordering. If package A Requires B,
+then provider of B will be installed before A, and erased after it.
+Weak dependencies (`Recommends`, `Suggests, `Supplements, `Enhances`)
```
Supplements is missing a closing back tick.

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+# Package ordering
+
+Rpm orders transactions solely based on package dependency information:
+dependencies first on install, last on erase. If the install/erase
```
I find this a bit hard to read.  Perhaps: ...dependencies are ordered first on install...

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+
+A dependency loop is generally a packaging bug as no correct ordering is
+possible to achieve in that situation. They are best solved by splitting
+to sub-packages to eliminate the loop, but sometimes this is not practical,
```
... splitting *a package* into sub-packages...

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+to sub-packages to eliminate the loop, but sometimes this is not practical,
+and also some dependencies are not relevant for ordering at all. Rpm
+supports various [dependency qualifiers](spec.md#Requires) to supply
+context which rpm can and does use to improve ordering and resolving
```
s/which/that/

pmatilai replied:
Oh. I had to look it up, and if the difference has come up in my English lessons I've completely forgotten it in the intervening 30+ years. There's a very similar word pair "joka" and "mikä" in Finnish whose (rather common) misuse drives me crazy, thanks for the lesson :grinning:

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+
+Normally, `Requires: A` and `Requires(pre): A` result in exactly the same
+order but when a loop is detected, a dependency with an ordering-qualifier
+are favored.
```
s/are/is/

ffesti replied:
expect for weak dependencies which are always treated as without qualifiers.

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+are favored.
+
+Both install and erase ordering qualifiers are additionally used as loop
+cutting hints in presence of dependency loops. Packages belonging to a
```
*the* presence

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+
+Both install and erase ordering qualifiers are additionally used as loop
+cutting hints in presence of dependency loops. Packages belonging to a
+dependency loop are installed as a consequtive group, using the qualifier
```
consecutive

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+
+Ordering issues and dependency loops can be difficult to analyze,
+to aid with this rpm can print out loop diagnostics from transactions.
+The diagnostics can be enabled with `--deploops` switch from the cli
```
with *the*

nwalfield commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+Ordering issues and dependency loops can be difficult to analyze,
+to aid with this rpm can print out loop diagnostics from transactions.
+The diagnostics can be enabled with `--deploops` switch from the cli
+or `RPMTRANS_FLAG_DEPLOOPS` transaction flag from the API. Installing
```
or *the*...

nwalfield reviewed (commented):
The text seems internally consistent, but I'm not familiar enough with how rpm works to check most of the technical bits.

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+context which rpm can and does use to improve ordering and resolving
+dependency loops.
+
+Qualifiers affecting install ordering:
```
"only" needed here , too. May be a bit longer sentence made things clearer "Dependencies wit these qualifiers affect install ordering only"

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+order but when a loop is detected, a dependency with an ordering-qualifier
+are favored.
+
+Both install and erase ordering qualifiers are additionally used as loop
```
This paragraph and the one above are more or less the same and should be merged into one.

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+hints to try and cut the loop with least minimal disruption, but one should
+remember that there is no such thing as a correct order for a loop.
+
+Qualifiers denoting an unordered dependency, ie dismissing a dependency from
```
Too complicated language! May be "Dependencies with the following qualifiers are ignored for ordering:"

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+* `verify`
+
+Ordering issues and dependency loops can be difficult to analyze,
+to aid with this rpm can print out loop diagnostics from transactions.
```
analyze. To

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,56 @@
+order is not right, then either the dependency information is incomplete,
+incorrect or there are unresolvable dependency loops.
+
+`Requires` are the primary data for ordering. If package A Requires B,
```
Not sure if "primary data" makes a lot of sense here. May be the match between a Dependency like Requires and a Provide or Name of another should be mentioned here instead. This would be the place to establish this  as a "Dependency" so the word has a defined meaning for the rest of the document.

ffesti reviewed (commented):
Overall I like this very much. It doesn't dive too much into the implementation details but still gives the relevant information to packagers.

pmatilai commented:
Thanks for the proof-reading and other very legit suggestions. I think I addressed all the issues pointed out, but may have of course introduced new ones in the process as some sections needed more than just cosmetics.

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,69 @@
+is only used on hard `Requires` dependencies.
+
+`Requires(preun): A` declares the dependency is only needed during erasure,
+so it can resolve an install-time loop. Unordering qualifiers cause the
```
May be "non-ordering" is a better term here? May be a native speaker can chime in on that.

pmatilai replied:
I chose unorder intentionally as those qualifiers *unorder* the dependency. Names are hard, native speakers input welcome of course.

nwalfield replied:
The fundamental problem is that rpm needs to build a DAG (partial order) out of the dependencies.  But to create a DAG, there can't be circular dependencies (i.e., loops).  To create a DAG, rpm breaks loops by removing edges.  Because removing an edge means that a package will be installed (or removed) before a package that it depends on is installed, this is fragile.  rpm provides a mechanisms to influence how it should break these dependency loops.

Given the above: instead of using the term unordering qualifier, perhaps dependency refinement would be better.

pmatilai replied:
We use the term qualifier for the "stuff between the parentheses" in other places, so I think it makes sense to stick to the terminology even if less than perfect.

ffesti commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,69 @@
+correct order for a loop.
+
+Ordering issues and dependency loops can be difficult to analyze.
+To aid with this rpm can print out loop diagnostics from transactions.
```
for transactions ?

pmatilai replied:
Dunno. To me it seems the diagnostics are coming from (inside) transactions, not made for them. Native speaker wanted :sweat_smile:

nwalfield replied:
A third option :sweat_smile:: about transactions

pmatilai replied:
Routed around the problem by rewriting: "To aid with this, rpm can print out transaction ordering diagnostics."

pmatilai commented on docs/manual/tsort.md:
```diff
@@ -2,160 +2,69 @@
+To aid with this rpm can print out loop diagnostics from transactions.
+These diagnostics can be enabled with the `--deploops` switch from the cli
+or the `RPMTRANS_FLAG_DEPLOOPS` transaction flag from the API. Installing
+to (and erasing from) an empty chroot is an effective way of discovering
```
Sorry but no, this is about explaining rpm ordering process, not a howto.

ffesti replied:
Fair enough.

pmatilai commented:
Okay, this isn't a book to be printed we're talking about, just another live doc in a git repo that we're free to revisit anytime anybody wants, no point getting stuck on details especially when the existing documentation is so terribly outdated. Thanks for the valuable feedback guys!
